node/node.go

```diff
@@ -132,6 +135,116 @@ func StateProvider(stateProvider statesync.StateProvider) Option {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	if config == nil {
+		logger.Info("no config provided, using default configuration")
+		config = cfg.DefaultConfig()
+	}
+
+	if dbProvider == nil {
+		dbProvider = cfg.DefaultDBProvider
+	}
+	blockStore, stateDB, err := initDBs(config, dbProvider)
+
+	if err != nil {
+		return err
+	}
+	if !blockStore.IsEmpty() {
+		return fmt.Errorf("blockstore not empty, trying to initialize non empty state")
+	}
+
+	stateStore := sm.NewStore(stateDB, sm.StoreOptions{
+		DiscardABCIResponses: config.Storage.DiscardABCIResponses,
+	})
+
+	state, err := stateStore.Load()
+	if err != nil {
+		return err
+	}
+
+	if !state.IsEmpty() {
+		return fmt.Errorf("state not empty, trying to initialize non empty state")
+	}
+
+	genState, _, err := LoadStateFromDBOrGenesisDocProvider(stateDB, DefaultGenesisDocProviderFunc(config))
+	if err != nil {
+		return err
+	}
+
+	stateProvider, err := statesync.NewLightClientStateProvider(
+		ctx,
+		genState.ChainID, genState.Version, genState.InitialHeight,
+		config.StateSync.RPCServers, light.TrustOptions{
+			Period: config.StateSync.TrustPeriod,
+			Height: config.StateSync.TrustHeight,
+			Hash:   config.StateSync.TrustHashBytes(),
+		}, logger.With("module", "light"))
+	if err != nil {
+		return fmt.Errorf("failed to set up light client state provider: %w", err)
+	}
+
+	state, err = stateProvider.State(ctx, height)
+	if err != nil {
+		return err
+	}
+	if appHash == nil {
+		logger.Info("warning: cannot verify appHash. Verification will happen when node boots up!")
+	} else {
+		if !bytes.Equal(appHash, state.AppHash) {
+			if err := blockStore.Close(); err != nil {
+				logger.Error("failed to close blockstore")
+			}
+			if err := stateStore.Close(); err != nil {
+				logger.Error("failed to close statestore")
+			}
+			return fmt.Errorf("the app hash returned by the light client does not match the provided appHash, expected %X, got %X", state.AppHash, appHash)
+		}
+	}
+
+	commit, err := stateProvider.Commit(ctx, height)
+	if err != nil {
+		return err
+	}
+
+	if err := stateStore.Bootstrap(state); err != nil {
+		return err
+	}
+
+	err = blockStore.SaveSeenCommit(state.LastBlockHeight, commit)
+	if err != nil {
+		return err
+	}
+
+	// Once the stores are bootstrapped, we need to set the height at which the node has finished
+	// statesyncing. This will allow the blocksync reactor to fetch blocks at a proper height.
+	// In case this operation fails, it is equivalent to a failure in  online state sync where the operator
+	// needs to manually delete the state and blockstores and rerun the bootstrapping process.
+	err = stateStore.SetOfflineStateSyncHeight(state.LastBlockHeight)
+	if err != nil {
+		return fmt.Errorf("failed to set synced height")
+	}
+	if err = blockStore.Close(); err != nil {
+		logger.Error("failed to close blockstore")
+	}
+	var err2 error
+	if err2 = stateStore.Close(); err2 != nil {
+		logger.Error("failed to close statestore")
+		return err2
+	}
```

Comment:
It looks to me as though all these manual close operations spread throughout this function can simply be replaced by the `defer` statements in my other two comments.

In general it's better to do this sort of thing in a `defer` because, even if the function panics somewhere, the `defer` statements will still be called. In the current version of this function, that won't happen, and it's not clear to me what the impact on the stores would be as a result.

Reply:
Yes, if they are not closed properly then re-opening them will be a problem. The only problem is what if close fails. But we can have another function that tries to close them, returning an error and call this function with defer. 

`defer func() (error) {

if err = blockStore.Close(); err != nil {
		logger.Error("failed to close blockstore")
	}
var err2 error
if err2 = stateStore.Close(); err2 != nil {
		logger.Error("failed to close statestore")
		return err2
}
return err
}
`

Reply:
It's always best to separate each close operation into separate `defer` statements. Also, the `(error)` return value for the function would end up being ignored. If the enclosing function doesn't need to know about the errors, one could simply:

```go
defer func() {
	if err := blockStore.Close(); err != nil {
		logger.Error("Failed to close blockstore", "err", err)
	}
}()
// ...
defer func() {
	if err := stateStore.Close(); err != nil {
		logger.Error("Failed to close statestore", "err", err)
	}
}()
```

If the enclosing function _does_, however, need to return any errors produced from within the `defer` statements, you'd have to name the return value in the function signature and use it directly in the `defer` statements:

```go
func BootstrapState(
	ctx context.Context,
	config *cfg.Config,
	dbProvider cfg.DBProvider,
	height uint64,
	appHash []byte,
) (err error) {
	// ...
	defer func() {
		if derr := blockStore.Close(); derr != nil {
			logger.Error("Failed to close blockstore", "err", derr)
			// Set the return value
			err = derr
		}
	}()

	// ...
	defer func() {
		if derr := stateStore.Close(); derr != nil {
			logger.Error("Failed to close statestore", "err", derr)
			err = derr
		}
	}()

	// ...
}
```

Both `defer` statements will be executed, and whichever of them errors last will influence the actual function result.

Reply:
Thanks, will refactor this then. I think we should return errors so I'll take the second approach.

Reply:
@thanethomson is the current way of closing the stores ok? So we can merge this PR:)